From 6c076b4ea9a963249987daa158808e33910e86e3 Mon Sep 17 00:00:00 2001
From: Lior Suliman <liorsu@gmail.com>
Date: Tue, 2 Jun 2026 14:00:46 +0000
Subject: [PATCH] Fix buffer overflow and missing null-termination in name2id

When idmapping is disabled, `atless2id` was called directly with
`name->addr` (which is not guaranteed to be null-terminated), risking
out-of-bounds reads. Fix this by copying `name->addr` to a stack-
allocated, null-terminated buffer (`namebuff`) before the call.

Change-Id: I944a8cc399da9b269ef11ff4319b63b8bb88f391
Signed-off-by: Lior Suliman <liorsu@gmail.com>
---
 src/Protocols/NFS/nfs_proto_tools.c | 28 ++++++++++++++++++++++------
 src/idmapper/idmapper.c             |  4 ++--
 src/include/nfsv41.h                |  2 +-
 3 files changed, 25 insertions(+), 9 deletions(-)

diff --git a/src/Protocols/NFS/nfs_proto_tools.c b/src/Protocols/NFS/nfs_proto_tools.c
index efb79bd619..4c0fd8a0a4 100644
--- a/src/Protocols/NFS/nfs_proto_tools.c
+++ b/src/Protocols/NFS/nfs_proto_tools.c
@@ -1661,6 +1661,8 @@ static fattr_xdr_result decode_owner(XDR *xdr, struct xdr_attrs_args *args)
 	uid_t uid;
 	uint32_t len = 0;
 	struct gsh_buffdesc ownerdesc;
+	void *src_owner;
+	char *owner_buff;
 	unsigned int pos, newpos;
 
 	if (!inline_xdr_u_int(xdr, &len))
@@ -1676,15 +1678,21 @@ static fattr_xdr_result decode_owner(XDR *xdr, struct xdr_attrs_args *args)
 	if (len % 4 != 0)
 		newpos += (4 - (len % 4));
 
-	ownerdesc.len = len;
-	ownerdesc.addr = xdr_inline_decode(xdr, len);
+	src_owner = xdr_inline_decode(xdr, len);
 
-	if (!ownerdesc.addr) {
+	if (!src_owner) {
 		LogMajor(COMPONENT_NFS_V4,
 			 "xdr_inline_decode on xdrmem stream failed!");
 		return FATTR_XDR_FAILED;
 	}
 
+	/* Allocate a temporary null terminated buffer */
+	owner_buff = alloca(len + 1);
+	memcpy(owner_buff, src_owner, len);
+	owner_buff[len] = '\0';
+	ownerdesc.len = len;
+	ownerdesc.addr = owner_buff;
+
 	if (!name2uid(&ownerdesc, &uid, get_anonymous_uid())) {
 		args->nfs_status = NFS4ERR_BADOWNER;
 		return FATTR_BADOWNER;
@@ -1711,6 +1719,8 @@ static fattr_xdr_result decode_group(XDR *xdr, struct xdr_attrs_args *args)
 	gid_t gid;
 	uint32_t len = 0;
 	struct gsh_buffdesc groupdesc;
+	void *src_group;
+	char *group_buff;
 	unsigned int pos, newpos;
 
 	if (!inline_xdr_u_int(xdr, &len))
@@ -1726,15 +1736,21 @@ static fattr_xdr_result decode_group(XDR *xdr, struct xdr_attrs_args *args)
 	if (len % 4 != 0)
 		newpos += (4 - (len % 4));
 
-	groupdesc.len = len;
-	groupdesc.addr = xdr_inline_decode(xdr, len);
+	src_group = xdr_inline_decode(xdr, len);
 
-	if (!groupdesc.addr) {
+	if (!src_group) {
 		LogMajor(COMPONENT_NFS_V4,
 			 "xdr_inline_decode on xdrmem stream failed!");
 		return FATTR_XDR_FAILED;
 	}
 
+	/* Allocate a temporary null terminated buffer */
+	group_buff = alloca(len + 1);
+	memcpy(group_buff, src_group, len);
+	group_buff[len] = '\0';
+	groupdesc.len = len;
+	groupdesc.addr = src_group;
+
 	if (!name2gid(&groupdesc, &gid, get_anonymous_gid())) {
 		args->nfs_status = NFS4ERR_BADOWNER;
 		return FATTR_BADOWNER;
diff --git a/src/idmapper/idmapper.c b/src/idmapper/idmapper.c
index 5cf5787689..ddac896fc4 100644
--- a/src/idmapper/idmapper.c
+++ b/src/idmapper/idmapper.c
@@ -1109,7 +1109,6 @@ static bool name2id(const struct gsh_buffdesc *name, uint32_t *id, bool group,
 
 	if (!idmapping_enabled) {
 		/* If idmapping isn't enabled we accept numeric ids only */
-		/* Something we can mutate and count on as terminated */
 		return atless2id(name->addr, name->len, id, anon);
 	}
 
@@ -1142,7 +1141,8 @@ static bool name2id(const struct gsh_buffdesc *name, uint32_t *id, bool group,
 
 	/* Something we can mutate and count on as terminated */
 	namebuff = alloca(name->len + 1);
-	memcpy(namebuff, name->addr, name->len + 1);
+	memcpy(namebuff, name->addr, name->len);
+	*(namebuff + name->len) = '\0';
 	at = memchr(namebuff, '@', name->len);
 
 	if (at == NULL) {
diff --git a/src/include/nfsv41.h b/src/include/nfsv41.h
index ce38058b7d..4abe7233a1 100644
--- a/src/include/nfsv41.h
+++ b/src/include/nfsv41.h
@@ -9776,7 +9776,7 @@ static inline bool xdr_CB_COMPOUND4res(XDR *xdrs, CB_COMPOUND4res *objp)
 static inline void copy_into_utf8string(utf8string *dest, const char *src,
 					int len)
 {
-	dest->utf8string_val = gsh_malloc(len + 1);
+	dest->utf8string_val = (char *)gsh_malloc(len + 1);
 	dest->utf8string_len = len;
 	memcpy(dest->utf8string_val, src, len);
 	dest->utf8string_val[len] = '\0';
